core/base/src/main/java/alluxio/collections/ConcurrentIdentityHashMap.java

```diff
@@ -0,0 +1,383 @@
+   * elements is negative or the load factor is nonpositive
+   */
+  public ConcurrentIdentityHashMap(int initialCapacity, float loadFactor) {
+    this(initialCapacity, loadFactor, 1);
+  }
+
+  /**
+   * Creates a new, empty map with an initial table size based on the given number of elements
+   * ({@code initialCapacity}), load factor ({@code loadFactor}), and number of
+   * expected concurrently updating threads ({@code concurrencyLevel}).
+   *
+   * @param initialCapacity the initial capacity. The implementation performs internal sizing to
+   *        accommodate this many elements, given the specified load factor.
+   * @param loadFactor the load factor (table density) for
+   *        establishing the initial table size
+   * @param concurrencyLevel the estimated number of concurrently updating threads. The
+   *        implementation may use this value as a sizing hint.
+   * @throws IllegalArgumentException if the initial capacity is negative or the load factor or
+   *         concurrencyLevel are nonpositive
+   */
+  public ConcurrentIdentityHashMap(int initialCapacity,
+      float loadFactor, int concurrencyLevel) {
+    mInternalMap = new ConcurrentHashMap<>(initialCapacity, loadFactor, concurrencyLevel);
+  }
+
+  @Override
+  public boolean containsKey(Object k) {
+    return mInternalMap.containsKey(new IdentityObject<>(k));
+  }
+
+  @Override
+  public V put(K k, V v) {
+    return mInternalMap.put(new IdentityObject<>(k), v);
+  }
+
+  @Override
+  public V putIfAbsent(K k, V v) {
+    return mInternalMap.putIfAbsent(new IdentityObject<>(k), v);
+  }
+
+  @Override
+  public boolean remove(Object k, Object v) {
+    return mInternalMap.remove(new IdentityObject<>(k), v);
+  }
+
+  @Override
+  public V remove(Object k) {
+    return mInternalMap.remove(new IdentityObject<>(k));
+  }
+
+  @Override
+  public boolean replace(K k, V v1, V v2) {
+    return mInternalMap.replace(new IdentityObject<>(k), v1, v2);
+  }
+
+  @Override
+  public V replace(K k, V v) {
+    return mInternalMap.replace(new IdentityObject<>(k), v);
+  }
+
+  @Override
+  public V get(Object k) {
+    return mInternalMap.get(new IdentityObject<>(k));
+  }
+
+  @Override
+  public Collection<V> values() {
+    return mInternalMap.values();
+  }
+
+  @Override
+  public void clear() {
+    mInternalMap.clear();
+  }
+
+  /**
+   * Returns a set representing the entries of this map at the time of calling.
+   *
+   * Note, this method will create a copy of the map's entry set at creation time. Standard
+   * behavior in the {@link ConcurrentHashMap} says that the entry set backed by the map is
+   * updated when the underlying map is updated. That is not the case for this class.
+   *
+   * @return The set of entries in the map at time of calling
+   */
+  @Override
+  public Set<Map.Entry<K, V>> entrySet() {
+    return mInternalMap.entrySet().stream().map(IdentityEntry<K, V>::new)
+        .collect(Collectors.toSet());
```

Comment:
Will this be much more expensive, why can't we use the default implementation?

Reply:
JFYI, we'll only be using `put`, `remove` and `values` methods.

Reply:
The internal map's entry set returns type `Set<Map.Entry<IdentityObject<K>, V>>`. To have a compatible return type I needed to this. Unfortunately I couldn't find a way to automatically unbox the `K` type from the `IdentityObject<K>` so I settled with this slower implementation for now. The caveats are noted in the javadoc of the class and on the method.


I think it would be possible to implement an auto-unboxing class internally, but I felt that because we don't use `entrySet()` elsewhere in the codebase that this would be unnecessary work. If you feel having an auto-unboxing implementation is necessary, then I can probably add one (similar to what I did for `keySet()`) but it will add a good bit of implementation

Reply:
Thanks for the explanation.